Add browser support

mateonunez opened the pull request:
This PR would introduce the browser support for `async-cache-dedupe` as discussed in #47.

## Context

This library is designed to run on Node.js. With support for Chrome, Firefox, Safari, and Edge., the `async-cache-dedupe` module can be extended between platforms and, in the near future, to other runtimes.

- [x] Test Runner.
- [x] Browser runner.
- [x] Prepare runner.
- [x] ESBuild.
- [x] Rollup.
- [x] Webpack.
- [x] Browserify.
- [x] Vite.
- [x] Chrome.
- [x] Firefox.
- [x] Safari.
- [x] Edge.
- [x] Tests. (`util` was not tested)
- [x] Documentation.
- [x] CI.

## Test

I've added a test system to run all tests in the following browsers: Chrome, Safari, Edge, and Firefox. I also add support for the following bundlers: Esbuild, Webpack, Browserify, and Rollup.

To test correctly each bundler you need to run the following commands:

```bash
npm run test:prepare [esbuild|rollup|browserify|webpack]
```

This command will generate a JS file bundled and move it into a temporary folder.

To run the tests you should run the following command:

```bash
npm run test:browser [chrome|firefox|safari|edge]
````

This gonna run an `index.html` file that imports the generated suites containing tests. All the output will be streamed from the console of the browser to your Node console.

## CI

The CI was added to run the browser tests in the following OS: `ubuntu`, `windows`, and `macos`. Each OS tests all the supported browsers with the supported bundlers. I've skipped the **Safari** tests on **Ubuntu** and **Windows**.

## Caveats

There are some limitations to running this module on the client side.

- Redis Storage will not available on the client side. (throws an exception)

mcollina commented:
Good work! I think we might want to support Vite too.

I think we should not try to bundle https://github.com/mcollina/async-cache-dedupe/blob/main/src/storage/redis.js at all.

mateonunez commented:
Hey @mcollina, thanks for the pre-review 😁

I've added `Vite` support as well and removed `redis.js` from bundles. The PR is still in draft because it fails randomly. I need to investigate the error. I think it's because of timers, setImmediate, or now method. I'll let you know as soon as possible.

mcollina reviewed (changes requested):
Can you just remove the skipped tests?

mcollina reviewed (commented):
A lot of tests here are duplication of the Node.js ones. Could they be imported/loaded instead of copying them over?

mateonunez commented:
> A lot of tests here are duplication of the Node.js ones. Could they be imported/loaded instead of copying them over?

Oh, I've tried that way... but I realized there are no shims for `async_hooks` (used by tap).

mcollina reviewed (approved):
lgtm

mcollina commented:
let's see what @simone-sanfratello says!

simone-sanfratello commented on package.json:
```diff
@@ -6,6 +6,8 @@
   "types": "index.d.ts",
   "scripts": {
     "test": "standard | snazzy && tap test/*test.js && tsd",
+    "test:prepare": "node test/browser/helpers/runner-prepare.mjs",
```
rename "test:prepare" to "pretest:browser" ?

mateonunez replied:
Unfortunately, script hooks have a limitation and that is that it is unable to forward arguments to the main script.

One solution might be to create another file that executes both: the prepare and the runner. At this point we will only have one script, it also seems to me a much cleaner solution. 

What do you think?

simone-sanfratello replied:
Is it possible to include the `setup` script (let's use `setup` to identify pre* operations) in the tests, and check on there if the setup phase is done?

simone-sanfratello commented on README.md:
```diff
@@ -311,6 +311,40 @@ const p1 = cache.fetchSomething(42) // <--- TypeScript doesn't argue anymore her
 
+## Browser
+
+This module is designed to work in a browser environment, and is fully compatible with all major browsers. However, it does not support the use of `Redis` as a store. Instead, it utilizes the memory to cache and store the data.
```
```
All the major browser are supported; only `memory` storage type is supported, `redis` storage can't be used in a browser env.
```

simone-sanfratello commented on package.json:
```diff
@@ -30,15 +32,34 @@
+    "webpack-cli": "^5.0.1"
   },
   "dependencies": {
-    "abstract-logging": "^2.0.1",
```
missing dep

simone-sanfratello replied:
please check deps

mateonunez replied:
Oh, I remove it since is not compatible with browsers. I wrote a simple function to simulate the abstractLogging and I've reused it in both envs.

Should I restore it and keeping the function for browsers?

simone-sanfratello replied:
Oh, I haven't noticed, but I prefer to keep the `abstract-logging` for browser too, I think it can work on browser too

mateonunez replied:
You're absolutely right. It's working on browsers too. Fixing it.

simone-sanfratello commented:
Nice work! Looks fine so far.
Please remove all the `// istanbul ignore` and provide meaningful tests

simone-sanfratello commented:
All good with the latest changes! :clap: 

One last thing please: remove `test:prepare` script and run it in `test:browser` if needed

mateonunez commented:
Hey, @simone-sanfratello. Thank you so much for the review and for your support. I've made all changes you requested.

simone-sanfratello commented:
Well done!
